linalg/libBatchSolver.hpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_LIB_BATCH_SOLVER
+#define MFEM_LIB_BATCH_SOLVER
+
+#include "../config/config.hpp"
+#include "../general/globals.hpp"
+#include "matrix.hpp"
+#include "densemat.hpp"
+
+#if defined(MFEM_USE_CUDA)
+// These macros prefixes the argument with `cuda` or `hip`, e.g. `cudaStream_t` or `hipStream_t`
+#define MFEM_SUB_cuda_or_hip(stub) cuda##stub
+#define MFEM_SUB_cu_or_hip(stub) cu##stub
+#define MFEM_SUB_Cuda_or_Hip(stub) Cuda##stub
+#define MFEM_SUB_CUDA_or_HIP(stub) CUDA##stub
+#define MFEM_SUB_CU_or_HIP(stub) CU##stub
+
+#include <cuda.h>
+#include <cublas.h>
+#include <cusolverDn.h>
+#include <cuda_runtime.h>
+#endif  // MFEM_USE_CUDA
+
+#if defined(MFEM_USE_HIP)
+// These macros prefixes the argument with `cuda` or `hip`, e.g. `cudaStream_t` or `hipStream_t`
+#define MFEM_SUB_cu_or_hip(stub) hip##stub
+#define MFEM_SUB_cuda_or_hip(stub) hip##stub
+#define MFEM_SUB_Cuda_or_Hip(stub) Hip##stub
+#define MFEM_SUB_CUDA_or_HIP(stub) HIP##stub
+#define MFEM_SUB_CU_or_HIP(stub) HIP##stub
```

Comment:
since we need these in the header here and we have them in https://github.com/mfem/mfem/blob/master/linalg/sparsemat.cpp can we move them to https://github.com/mfem/mfem/blob/master/general/cuda.hpp and https://github.com/mfem/mfem/blob/master/general/hip.hpp to avoid duplicates/redefined symbols?